src/compiler/WebSharper.FSharp.Service/Program.fs

```diff
@@ -30,7 +31,50 @@ open WebSharper.Compiler
 open WebSharper.Compiler.FSharp.Compile
 open WebSharper.Compiler.FSharp.ErrorPrinting
 open WebSharper.Compiler.CommandTools
+open System.Collections.Generic
 
+type FileTimestamp = { Path: string; Timestamp: DateTime }
+type CachedProjInfo = { Timestamps: FileTimestamp list; Args: string [] }
+let argsDict = Dictionary<string, CachedProjInfo>(StringComparer.InvariantCultureIgnoreCase)
+
+let (|Exit|FullCompile|PostCompile|) (args: ArgsType) = 
+    let tryGetDirectoryName (path: string) =
+        try
+            let path = System.IO.Path.GetDirectoryName path
+            if System.IO.Directory.Exists path then
+                Some path
+            else
+                None
+        with
+        | _ -> None
+
+    if args = {args= [|"exit"|]} then
+        Exit
+    elif args.args.Length = 1 && args.args.[0].StartsWith("compile:") then
+        let projectOption = args.args.[0].Substring(8) 
+        let projectDirOption = projectOption |> Some |> Option.bind tryGetDirectoryName
+        match projectDirOption with
+        | Some project -> System.Environment.CurrentDirectory <- project
```

Comment:
I would move this outside of active pattern, better to keep it side effect-free.

Reply:
Applied the suggested changes